philippemnoel/whist
# wcc with trendline filter / delay-base controller

## backgroud
See gist: https://gist.github.com/wangyu-/e6208baf47945d56595872d9af5c6542
## Description

New version of WCC with delay-based controller from GCC correctly implemented, 

Addresses issues of https://github.com/whisthq/whist/issues/7380

## Implementation

## Documentation & Tests Added

## Testing Instructions

the version avaliable for test is cf244a289914f8edb060a29fff5d73d0ab1a59b0, which has been rebased to the lasted dev.

If you run with `--wcc_v2 --always-saturate` on client side,  then wcc_v2 is running.  (at the moment, bandwitdh saving logic for wcc_v2 is not implement, always-saturate is needed)

If you run with no additional parameter, then the original wcc is running.   

you can also run with `--always-saturate`,   then it's original wcc with bandwith saving logic off. So that you can compare wcc and wcc_v2 without the affected of decoder undershooting.

No matter wcc or wcc_v2, the plot data will be automatically generated at `/tmp/plot.json`, if you encounter anything weird, sent the plot file to me.

wcc and wcc_v2 can share the same server, you don't need to switch between wcc and wcc_v2 on server side. (but make sure your server is also on cf244a289914f8edb060a29fff5d73d0ab1a59b0) 

### What to expected

#### On a very good  link  (with very large bandwith, no competing traffic)

there shouldn't be much different with the old wcc, since for such a good link congestion control is not really taking into account.

#### On a constrained link (either the link's bandwith is small,  or there are competing traffic, or the channel itself is noisy)

1. less freeze time on first connection.  i.e. you might encountered a lot, after you start whist client, the video got freezed for a few seconds.   This will get reduced.   (in future the "prober" idea from webrtc can hopefully solve this completely.  )

2. less latency spikes.  which in further means, less audio pops, less video freezes (tiny or big),  less response time for click or play online game.   If the difference is too subtle to tell, you can see the difference from `short_term_latency` from the plot.

3. less weird loss of connection.

drawback: 

1. a little less clearness/sharpness of text sometime.   Since anyway congestion void is achieved by reduce birate.  But make sure this doesn't happen too often or too much.

### Testing suggestions

1. pay more attention to test on real world network (for example, office wifi, office wired network), instead of simulated ideal link.

Since I have tested on simulated environments  like a thousand times, there shouldn't be big problems.    The real world networks are the parts my test can't could cover.

2. pay more attention to the clearness/sharpenss of video,  instead of trying to feel the difference of latency.

the clearness/sharpenss  on real world network is what I am worrying most.   In term of latency, wcc_v2 should be almost always better (or not worse).  The worry is,  to achieve better latency, if video quailty is compromised (too much).

4. when comparing wcc and wcc_v2, especially for this version, it's prefered to test on moving content (like video, html5 animation , online game) than mostly static content.

Since on static content, for original wcc (with `--always-saturate` off),  the encoder is in serious undershoot most of the time, the congestion control is not taking too much account.   


### Others
If you want to test on simulated links, [here](https://github.com/whisthq/whist/wiki/tc_scripts) is some scripts I use for simulated links.



## PR Checklist**

- [ ] Did the PR author fully test this PR end-to-end?
- [ ] Did one PR reviewer fully test this PR end-to-end?
- [ ] Did one PR reviewer conduct a thorough code design review?
